Ensure CompositeByteBuf.addComponent* handles buffer in consistent way a...

normanmaurer opened the pull request:
...nd not causes leaks

Motivation:

At the moment we have two problems:
 - CompositeByteBuf.addComponent(...) will not add the supplied buffer to the CompositeByteBuf if its empty, which means it will not be released on CompositeByteBuf.release() call. This is a problem as a user will expect everything added will be released (the user not know we not added it).
 - CompositeByteBuf.addComponents(...) will either add no buffers if none is readable and so has the same problem as addComponent(...) or directly release the ByteBuf if at least one ByteBuf is readable. Again this gives inconsistent handling and may lead to memory leaks.

Modifications:

 - Always add the buffer to the CompositeByteBuf and so release it on release call.

Result:

Consistent handling and no buffer leaks.

normanmaurer commented:
@trustin please review buddy!

fredericBregier commented:
@normanmaurer So you're back with this fix ;-)

Have you consider to release the empty added ByteBuf, as the newly empty ones (once read) instead of simply added them ? 
In various codec, this is used to cumulate ByteBuf as long as they are not full enough (considering the decoder) and then read once possible... Once readed it might be nice to release them in order to not keep them into the memory for nothing... I may feel it wrongly, but I guess that this could have side effect on memory, right ?
Doing this way will keep the contract (add a new buffer, and keep them as long as they are useful - readable - then releasing them)...
WDYT? (again, I may be wrong)

trustin commented:
Because the empty buffer being added will not affect the component offsets:

* We can add the empty buffer anywhere in the component list.
* We don't need to call `updateComponentOffsets(...)`.
* Some empty buffers are often unreleasable (e.g. `EmptyByteBuf`).

we can do a better job when we meet an empty buffer, rather than handling it the same way with a non-empty buffer.

normanmaurer commented:
@trustin so basically you say we should always "append" the empty buffer and not call updateComponentOffset ?

trustin commented:
Also, I want to make sure that an empty buffer in the middle of the component list does not mess up `findComponent()`. Please add a specific test case that make sure it never selects an empty component.

trustin commented:
@normanmaurer Yeah. Alternatively, we could simply keep the lazy list of the empty buffers separately to keep things simple.

normanmaurer commented:
@trustin that may be even better..

normanmaurer commented:
@trustin one drawback when we use a separate list would be that component(...) would not be consistent with what you add to the CompositeByteBuf. Not sure if it's a big deal, just not consistent. WDYT ?

normanmaurer commented:
@trustin ping

normanmaurer commented:
@trustin any comments to my comment ?

trustin commented:
Sorry that I'm late. I agree with you that it's not a big deal. A user can always use `componentAtOffset()`. Just make sure to document that an empty component will not be counted.

normanmaurer commented:
@trustin please review again.... 

Basically I only improved up to some point, as this allows us to ensure all the index/offset based methods work as expected when you add an empty buffer. I think this is a good tradeoff as adding empty buffers is kind of an exception anyway. 

As a side note we may be able to improve it more and pool CompositeByteBuf in PooledByteBufAllocator. I will come up with an extra PR.

trustin commented on buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java:
```diff
@@ -350,8 +333,13 @@ private void updateComponentOffsets(int cIndex) {
-        components.remove(cIndex).freeIfNecessary();
-        updateComponentOffsets(cIndex);
+        Component comp = components.remove(cIndex);
+        int len = comp.length;
```
Perhaps it's unnecessary to introduce a local variable?

normanmaurer replied:
maybe... I just wanted to ensure we not break something once freeIfNecessary may reset all things in Component later.

trustin commented:
Looks good to me except for one nit above. Please feel free to cherry-pick.

normanmaurer commented:
Addressed comments and cherry-picked into 4.0 (e2c8ae9e1e3dde4aceada5ddebf106abf07db423) , 4.1 (41fd857a7c749257c862ffb368af0b926c72ebfe) and master (b1cd4b545d5a83d2718786de72c80a032d7a1f3c)
